Update request: rnnoise 2021-01-22 → 0.2

- Package name: rnnoise
- Latest released version: 0.2
<!-- Search your package here: https://search.nixos.org/packages?channel=unstable -->
- Current version on the unstable channel: 2021-01-22
- Current version on the stable/release channel: 2021-01-22
<!--
Type the name of your package and try to find an open pull request for the package
If you find an open pull request, you can review it!
There's a high chance that you'll have the new version right away while helping the community!
-->
- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

**Notify maintainers**

@nh2 

-----

Note for maintainers: Please tag this issue in your PR.

https://gitlab.xiph.org/xiph/rnnoise/-/tree/v0.2
https://github.com/xiph/rnnoise/releases/tag/v0.2

Tried building with

```diff
-stdenv.mkDerivation (rec {
+stdenv.mkDerivation (finalAttrs: {
   pname = "rnnoise";
-  version = "2021-01-22";
+  version = "0.2";
 
   src = fetchFromGitHub {
     owner = "xiph";
     repo = "rnnoise";
-    rev = "1cbdbcf1283499bbb2230a6b0f126eb9b236defd";
-    sha256 = "1y0rzgmvy8bf9a431garpm2w177s6ajgf79y5ymw4yb0pik57rwb";
+    rev = "refs/tags/v${finalAttrs.version}";
+    sha256 = "sha256-Qaf+0iOprq7ILRWNRkBjsniByctRa/lFVqiU5ZInF/Q=";
   };
```

but it got a missing file.

```
       > In file included from src/denoise.c:43:
       > src/rnn.h:31:10: fatal error: rnnoise_data.h: No such file or directory
       >    31 | #include "rnnoise_data.h"
       >       |          ^~~~~~~~~~~~~~~~
       > compilation terminated.
       > make[1]: *** [Makefile:749: src/dump_features-denoise.o] Error 1
       > make[1]: Leaving directory '/build/source'
       > make: *** [Makefile:491: all] Error 2
```

There’s also warning about not using SSE3/AVX/AVX2, but Nix only supports the bottom of the barrel, common x86 instructions, but I usually compile this one out with some flags.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

[^1]

[^1]: Please consider [giving up MS GitHub](https://sfconservancy.org/GiveUpGitHub/) or offering a non-proprietary, non-US-corporate-controlled mirror for this free software project. I wish to delete this Microsoft account in the future, but I need more projects like this to support alternative methods to send patches & contribute.

--- a/pkgs/development/libraries/rnnoise/default.nix
+++ b/pkgs/development/libraries/rnnoise/default.nix
@@ -1,12 +1,16 @@
-{ stdenv, lib, fetchurl, fetchzip, autoreconfHook, writeScript }:
+{ stdenv, lib, fetchurl, fetchzip, autoreconfHook, writeScript
+, modelUrl ? "", modelHash ? "" # Allow overriding the model URL and hash
+}:
 
-let 
+let
   modelVersionJSON = lib.importJSON ./model-version.json;
 
-  # Copy from https://gitlab.xiph.org/xiph/rnnoise/-/raw/v${finalAttrs.version}/model_version
+  # Copy from https://gitlab.xiph.org/xiph/rnnoise/-/raw/v${version}/model_version
   default_model_version = modelVersionJSON.version;
-  default_model_url = "https://media.xiph.org/rnnoise/models/rnnoise_data-${default_model_version}.tar.gz";
-  default_model_hash = modelVersionJSON.hash;
+
+  # Either use the default model or the one provided by package override
+  model_url = if (modelUrl == "") then "https://media.xiph.org/rnnoise/models/rnnoise_data-${default_model_version}.tar.gz" else modelUrl;
+  model_hash = if (modelHash == "") then modelVersionJSON.hash else modelHash;
 
 in stdenv.mkDerivation (finalAttrs: {
   pname = "rnnoise";
@@ -21,8 +25,8 @@ in stdenv.mkDerivation (finalAttrs: {
   };
 
   model = fetchurl {
-    url = default_model_url;
-    hash = default_model_hash;
+    url = model_url;
+    hash = model_hash;
   };
 
   postPatch = ''

--- a/pkgs/development/libraries/rnnoise/default.nix
+++ b/pkgs/development/libraries/rnnoise/default.nix
@@ -1,15 +1,27 @@
-{ stdenv, lib, fetchFromGitHub, autoreconfHook }:
+{ stdenv, lib, fetchFromGitLab, fetchurl, autoreconfHook }:
 
-stdenv.mkDerivation (rec {
+stdenv.mkDerivation (finalAttrs: {
   pname = "rnnoise";
-  version = "2021-01-22";
+  version = "0.2";
 
-  src = fetchFromGitHub {
+  src = fetchFromGitLab {
+    domain = "gitlab.xiph.org";
     owner = "xiph";
     repo = "rnnoise";
-    rev = "1cbdbcf1283499bbb2230a6b0f126eb9b236defd";
-    sha256 = "1y0rzgmvy8bf9a431garpm2w177s6ajgf79y5ymw4yb0pik57rwb";
+    rev = "v${finalAttrs.version}";
+    hash = "sha256-Qaf+0iOprq7ILRWNRkBjsniByctRa/lFVqiU5ZInF/Q=";
   };
+
+  # Copy from https://gitlab.xiph.org/xiph/rnnoise/-/raw/v${finalAttrs.version}/model_version
+  model_version = "0b50c45";
+  model = fetchurl {
+    url = "https://media.xiph.org/rnnoise/models/rnnoise_data-${finalAttrs.model_version}.tar.gz";
+    hash = "sha256-SsgcXAiE7EvVkHAmqq4WIJt7ds2df3GvWCCUovmPS0M=";
+  };
+
+  patchPhase = ''
+    tar xvomf ${finalAttrs.model}
+  '';
 
   nativeBuildInputs = [ autoreconfHook ];
 
@@ -17,12 +29,12 @@ stdenv.mkDerivation (rec {
     install -Dt $out/bin examples/.libs/rnnoise_demo
   '';
 
-  meta = with lib; {
+  meta = {
     description = "Recurrent neural network for audio noise reduction";
     homepage = "https://people.xiph.org/~jm/demo/rnnoise/";
-    license = licenses.bsd3;
-    maintainers = [ maintainers.nh2 ];
+    license = lib.licenses.bsd3;
+    maintainers = with lib.maintainers; [ nh2 ];
     mainProgram = "rnnoise_demo";
-    platforms = platforms.all;
+    platforms = lib.platforms.all;
   };
 })

--- a/pkgs/development/libraries/rnnoise/default.nix
+++ b/pkgs/development/libraries/rnnoise/default.nix
@@ -1,25 +1,31 @@
-{ stdenv, lib, fetchFromGitLab, fetchurl, autoreconfHook }:
+{ stdenv, lib, fetchurl, fetchzip, autoreconfHook, writeScript }:
 
-stdenv.mkDerivation (finalAttrs: {
+let 
+  modelVersionJSON = lib.importJSON ./model-version.json;
+
+  # Copy from https://gitlab.xiph.org/xiph/rnnoise/-/raw/v${finalAttrs.version}/model_version
+  default_model_version = modelVersionJSON.version;
+  default_model_url = "https://media.xiph.org/rnnoise/models/rnnoise_data-${default_model_version}.tar.gz";
+  default_model_hash = modelVersionJSON.hash;
+
+in stdenv.mkDerivation (finalAttrs: {
   pname = "rnnoise";
   version = "0.2";
 
-  src = fetchFromGitLab {
-    domain = "gitlab.xiph.org";
-    owner = "xiph";
-    repo = "rnnoise";
-    rev = "v${finalAttrs.version}";
+  src = fetchzip {
+    urls = [
+      "https://gitlab.xiph.org/xiph/rnnoise/-/archive/v${finalAttrs.version}/rnnoise-v${finalAttrs.version}.tar.gz"
+      "https://github.com/xiph/rnnoise/archive/v${finalAttrs.version}.tar.gz"
+    ];
     hash = "sha256-Qaf+0iOprq7ILRWNRkBjsniByctRa/lFVqiU5ZInF/Q=";
   };
 
-  # Copy from https://gitlab.xiph.org/xiph/rnnoise/-/raw/v${finalAttrs.version}/model_version
-  model_version = "0b50c45";
   model = fetchurl {
-    url = "https://media.xiph.org/rnnoise/models/rnnoise_data-${finalAttrs.model_version}.tar.gz";
-    hash = "sha256-SsgcXAiE7EvVkHAmqq4WIJt7ds2df3GvWCCUovmPS0M=";
+    url = default_model_url;
+    hash = default_model_hash;
   };
 
-  patchPhase = ''
+  postPatch = ''
     tar xvomf ${finalAttrs.model}
   '';
 
@@ -27,6 +33,32 @@ stdenv.mkDerivation (finalAttrs: {
 
   postInstall = ''
     install -Dt $out/bin examples/.libs/rnnoise_demo
+  '';
+
+  passthru.updateScript = writeScript "update-rnnoise.sh" ''
+    #!/usr/bin/env nix-shell
+    #!nix-shell -i bash -p curl jq common-updater-scripts nix nix-prefetch findutils moreutils
+
+    prefetch-sri() {
+        nix-prefetch-url "$1" | xargs nix hash to-sri --type sha256
+    }
+
+    res="$(curl ''${GITHUB_TOKEN:+-u ":$GITHUB_TOKEN"} \
+      -sL "https://api.github.com/repos/xiph/rnnoise/tags?per_page=1")"
+
+    version="$(echo $res | jq '.[0].name | split("v") | .[1]' --raw-output)"
+    update-source-version ${finalAttrs.pname} "$version" --ignore-same-hash
+
+    model_version=$(curl -sL "https://raw.githubusercontent.com/xiph/rnnoise/v$version/model_version")
+    model_url="https://media.xiph.org/rnnoise/models/rnnoise_data-$model_version.tar.gz"
+    model_hash="$(prefetch-sri $model_url)"
+
+    modelJson=pkgs/development/libraries/rnnoise/model-version.json
+
+    jq --arg version "$model_version" \
+        --arg hash "$model_hash" \
+        '.version = $version | .hash = $hash' \
+        "$modelJson" | sponge "$modelJson"
   '';
 
   meta = {

--- a/pkgs/development/libraries/rnnoise/default.nix
+++ b/pkgs/development/libraries/rnnoise/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, lib, fetchurl, fetchzip, autoreconfHook, writeScript
+{ stdenv, lib, fetchurl, fetchzip, autoreconfHook, writeScript, fetchpatch
 , modelUrl ? "", modelHash ? "" # Allow overriding the model URL and hash
 }:
 
@@ -23,6 +23,14 @@ in stdenv.mkDerivation (finalAttrs: {
     ];
     hash = "sha256-Qaf+0iOprq7ILRWNRkBjsniByctRa/lFVqiU5ZInF/Q=";
   };
+
+  patches = [
+    # remove when updating
+    (fetchpatch {
+      url = "https://github.com/xiph/rnnoise/commit/372f7b4b76cde4ca1ec4605353dd17898a99de38.patch";
+      hash = "sha256-Dzikb59hjVxd1XIEj/Je4evxtGORkaNcqE+zxOJMSvs=";
+    })
+  ];
 
   model = fetchurl {
     url = model_url;

--- a/pkgs/development/libraries/rnnoise/model-version.json
+++ b/pkgs/development/libraries/rnnoise/model-version.json
@@ -1,0 +1,4 @@
+{
+  "version": "0b50c45",
+  "hash": "sha256-SsgcXAiE7EvVkHAmqq4WIJt7ds2df3GvWCCUovmPS0M="
+}

